lib/scripting/objects/vpm-controller.ts

```diff
@@ -18,16 +19,294 @@
  */
 
 import { Player } from '../../game/player';
+import { logger } from '../../util/logger';
+import { downloadGameEntry, LoadedGameEntry } from './pinball-backend/rom-fetcher';
+import { Emulator } from './pinball-backend/wpc-emu';
+
+/**
+ * Implementation of the VISUAL PINMAME COM OBJECT PROPERTY/METHOD
+ *
+ * converted VBS will call this functions using bracket notation, like `Controller.Dip[0]` or `Controller.GameName`
+ */
 
 export class VpmController {
 
 	private readonly player: Player;
+	private emulator: Emulator;
+	private gameName: string;
+	private splashInfoLine: string;
+	private paused: boolean;
+	public readonly Dip: { [index: number]: number };
+	public readonly Switch: { [index: number]: number };
+	public readonly Lamp: { [index: number]: number };
+	public readonly Solenoid: { [index: number]: number };
+	public readonly GIString: { [index: number]: number };
+
+	//private gameRomInfoPromise: Promise<Response>;
 
 	constructor(player: Player) {
 		this.player = player;
+		this.gameName = '';
+		this.splashInfoLine = '';
+		this.paused = false;
+		this.emulator = new Emulator();
+		// TODO route this to the emu
+		this.Dip = this.createDipGetter();
+		this.Switch = this.createGetSetBooleanProxy('SWITCH',
+			(index) => this.emulator.getSwitchInput(index), (switchNr, value) => this.emulator.setSwitchInput(switchNr, value));
+		this.Lamp = this.createGetSetNumberProxy('LAMP',
+			(index) => this.emulator.getLampState(index), SET_NOP);
+		this.Solenoid = this.createGetSetNumberProxy('SOLENOID',
+			(index) => this.emulator.getSolenoidState(index), SET_NOP);
+		this.GIString = this.createGetSetNumberProxy('GI',
+			(index) => this.emulator.getGIState(index), SET_NOP);
+		//		this.gameRomInfoPromise = Promise.reject();
+	}
+
+	// Control
+	get GameName(): string {
+		return this.gameName;
+	}
+	set GameName(gameName: string) {
+		this.gameName = gameName;
+		logger().debug('GAMENAME:', gameName);
+		//TODO this is hardwired to test
+		downloadGameEntry(gameName)
+			.then((answer: LoadedGameEntry) => {
+				logger().info('LOADED', answer.wpcDbEntry);
+				return this.emulator.loadGame(answer.wpcDbEntry, answer.romFile);
+			})
+			.then(() => {
+				this.player.setEmulator(this.emulator);
+			})
+			.catch((error) => {
+				logger().error('ERROR FAILED', error.messages);
+			});
```

Comment:
Can't make this async, but you can move it into an async function and just not await it here.

Reply:
will rework this anyway